src/Extensions/TimedScopes/Internal/ActivityObserversIntializer.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Omex.Extensions.Abstractions.Activities.Processing;
+
+namespace Microsoft.Omex.Extensions.TimedScopes
+{
+	internal sealed class ActivityObserversIntializer : IHostedService, IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object>>
+	{
+		private readonly IActivityStartObserver[] m_activityStartObservers;
+		private readonly IActivityStopObserver[] m_activityStopObservers;
+		private readonly LinkedList<IDisposable> m_disposables;
+		private IDisposable? m_observerLifetime;
+
+		public ActivityObserversIntializer(
+			IEnumerable<IActivityStartObserver> activityStartObservers,
+			IEnumerable<IActivityStopObserver> activityStopObservers)
+		{
+			m_activityStartObservers = activityStartObservers.ToArray();
+			m_activityStopObservers = activityStopObservers.ToArray();
+			m_disposables = new LinkedList<IDisposable>();
+		}
+
+		public Task StartAsync(CancellationToken cancellationToken)
+		{
+			m_observerLifetime = DiagnosticListener.AllListeners.Subscribe(this);
+			return Task.CompletedTask;
+		}
+		
+		public Task StopAsync(CancellationToken cancellationToken)
+		{
+			foreach (IDisposable disposable in m_disposables)
+			{
+				disposable.Dispose();
+			}
+			m_observerLifetime?.Dispose();
+			return Task.CompletedTask;
+		}
+
+		private bool IsActivityStart(string eventName) => m_activityStartObservers.Any() && eventName.EndsWith(".Start", StringComparison.Ordinal);
```

Comment:
>Any [](start = 77, length = 3)

`.Any()` uses LINQ which can be inefficient. Since we don't need LINQ here, it would be better to check `.Count != 0` (or similar).

Applies to the line below as well

Reply:
You may be able to remove the LINQ import if you remove this.

---
In reply to: [394383572](https://github.com/microsoft/Omex/pull/169#discussion_r394383572) [](ancestors = 394383572)

Reply:
Replaced, thanks, but I still need Linq to call `ToArray()` method